### ISSUE TITLE
Forward in U-net

### ISSUE THREAD

**Issue by royal-feng** [phillipi/pix2pix#65]
I want to do some change on middle feature of U-net , So I  seperated U-net to two part ,such as 1-to-8 layers called net-A. 9-to-53 layer called net-B. The process is, feature_middle = net-A:forward(image) then new_image = net-B:forward(feature_middle). 
But when it forward in net-B ,it failed. I think it because the jointable operation could not allow to seperated U-net to two part. So I want to ask how to reach my purpose on make some change on middle feature of U-net in another way?

the error is like it:
/home/zsf/torch/install/bin/luajit: /home/zsf/torch/install/share/lua/5.1/nn/Container.lua:67: 
In 18 module of nn.Sequential:
/home/zsf/torch/install/share/lua/5.1/nn/JoinTable.lua:24: 'for' limit must be a number
stack traceback:
	/home/zsf/torch/install/share/lua/5.1/nn/JoinTable.lua:24: in function </home/zsf/torch/install/share/lua/5.1/nn/JoinTable.lua:21>
	[C]: in function 'xpcall'
	/home/zsf/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/home/zsf/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	train.lua:255: in function 'createRealFake'
	train.lua:386: in main chunk
	[C]: in function 'dofile'
	.../zsf/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk

**Comment by Quasimondo** [comment]
Yes, there is a another way, at least if I understand correctly what you are trying to achieve. I made a version of the U-Net that allows me to element-wise multiply the values of the feature vector with a tensor before it gets passed on to the up-convolution part:

```
function defineG_unet(input_nc, output_nc, ngf)
    local netG = nil

    local f1 = nn.Identity()()

    -- input is (nc) x 256 x 256
    local e1 = - nn.SpatialConvolution(input_nc, ngf, 4, 4, 2, 2, 1, 1)
    -- input is (ngf) x 128 x 128
    local e2 = e1 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf, ngf * 2, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 2)
    -- input is (ngf * 2) x 64 x 64
    local e3 = e2 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 2, ngf * 4, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 4)
    -- input is (ngf * 4) x 32 x 32
    local e4 = e3 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 4, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8)
    -- input is (ngf * 8) x 16 x 16
    local e5 = e4 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 8, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8)
    -- input is (ngf * 8) x 8 x 8
    local e6 = e5 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 8, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8)
    -- input is (ngf * 8) x 4 x 4
    local e7 = e6 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 8, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8)
    -- input is (ngf * 8) x 2 x 2
    local e8 = e7 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 8, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8)
    -- input is (ngf * 8) x 1 x 1

    --these two lines were just my hackish way to figure out what dimensions the feature vector actually has, since using View gives an informative error message :-)
    --local dummy = e8 - nn.View(512,2,2) 
    --local mul = {dummy,f1} - nn.CMulTable(2)

    local mul = {e8,f1} - nn.CMulTable(2)

    local d1_ = mul - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 8, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8) - nn.Dropout(0.5)
    -- input is (ngf * 8) x 2 x 2
    local d1 = {d1_,e7} - nn.JoinTable(2)
    local d2_ = d1 - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 8 * 2, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8) - nn.Dropout(0.5)
    -- input is (ngf * 8) x 4 x 4
    local d2 = {d2_,e6} - nn.JoinTable(2)
    local d3_ = d2 - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 8 * 2, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8) - nn.Dropout(0.5)


    -- input is (ngf * 8) x 8 x 8
    local d3 = {d3_,e5} - nn.JoinTable(2)
    local d4_ = d3 - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 8 * 2, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8)
    -- input is (ngf * 8) x 16 x 16
    local d4 = {d4_,e4} - nn.JoinTable(2)
    local d5_ = d4 - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 8 * 2, ngf * 4, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 4)
    -- input is (ngf * 4) x 32 x 32
    local d5 = {d5_,e3} - nn.JoinTable(2)
    local d6_ = d5 - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 4 * 2, ngf * 2, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 2)
    -- input is (ngf * 2) x 64 x 64
    local d6 = {d6_,e2} - nn.JoinTable(2)
    local d7_ = d6 - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 2 * 2, ngf, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf)
    -- input is (ngf) x128 x 128
    local d7 = {d7_,e1} - nn.JoinTable(2)
    local d8 = d7 - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 2, output_nc, 4, 4, 2, 2, 1, 1)
    -- input is (nc) x 256 x 256

    local o1 = d8 - nn.Tanh()

    netG = nn.gModule({e1,f1},{o1})
    return netG
end
```
As you can see I've added another input f1, which is a tensor that has to have the same dimensions as the feature vectors. Those dimension change based on the batchSize, the fineSize and the number of filters used, so I calculate it this way:

`mask = torch.ones(opt.batchSize, opt.ngf/32*256,opt.fineSize / 256,opt.fineSize / 256)
`

When training you will now have to pass in two tensors, and have the mask filled with 1.0 so it does not have any influence. That part of the code looks like this:

```
if hasMask == true then
      fake_B = netG:forward({real_A,mask})
else
      fake_B = netG:forward(real_A)
end
```

I actually should have also added a second tensor which allows you to add values, that way you could replace parts or the entire feature vector with custom values. That part would use nn.CAddTable and come after the CMulTable, something like this:

```
local f2 = nn.Identity()()
{...}
local mul = {e8,f1} - nn.CMulTable(2)
local add = {mul,f2} - nn.CAddTable(2)

local d1 = add - nn.ReLU(true) - nn.SpatialFullConvolution(ngf * 8, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8) - nn.Dropout(0.5)

{...}

netG = nn.gModule({e1,f1,f2},{o1})
return netG
```

Another addition would be to also output the feature vector so you can analyze it, but I was lazy and solve that by iterating over the netG.forwardnodes:

```
local saveTensor = nil
if opt.saveTensors == 1 then
   local idx = 3
   for indexNode, node in ipairs(netG.forwardnodes) do
     local tn = torch.typename(node.data.module)
     if tn=='cudnn.SpatialConvolution' or tn=='cudnn.SpatialFullConvolution' then
       if idx==10 then
         saveTensor = node.data.module
         print ("saveTensor "..tn)
       end
       idx = idx + 1
     end
   end

end

```

**Comment by royal-feng** [comment]
@Quasimondo Thanks for your help. And you are indeed answer my confusing.
And I want to ask for more questions to you. 
1. I want to make more complex opertions on middle features, not only add or multiply. So I do it like this,


```
function defineG_unet(input_nc,output_nc,ngf)
  
local net_G=nil
  local e1 = - nn.SpatialConvolution(input_nc, ngf, 4, 4, 2, 2, 1, 1)
-- input is (ngf) x 128 x 128
    local e2 = e1 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf, ngf * 2, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 2)
    -- input is (ngf * 2) x 64 x 64
    local e3 = e2 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 2, ngf * 4, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 4) - swap
    -- input is (ngf * 4) x 32 x 32
  local e4 = e3 - nn.LeakyReLU(0.2, true) - nn.SpatialConvolution(ngf * 4, ngf * 8, 4, 4, 2, 2, 1, 1) - nn.SpatialBatchNormalization(ngf * 8)
{...}
```
and **swap** is a nn.module I defined some operation in it, but when I need to let image pass the netG, I need to update the swap because it related to another input such as f1. If I used two input like this  
`netG = nn.gModule({e1,f1},{o1})` ,maybe it's convenient.

2. You save the middle feature vector to realize analyze it .It's so cool.
In your code, 
```
local saveTensor = nil
if opt.saveTensors == 1 then
   local idx = 3
   for indexNode, node in ipairs(netG.forwardnodes) do
     local tn = torch.typename(node.data.module)
     if tn=='cudnn.SpatialConvolution' or tn=='cudnn.SpatialFullConvolution' then
       if idx==10 then
         saveTensor = node.data.module
         print ("saveTensor "..tn)
       end
       idx = idx + 1
     end
   end

end

```
indexNode, node are internal variable in ipairs(netG.forwardnodes)?How to use that code? insert that code after netG:forward(X)?
Thanks your so much!
```

**Comment by Quasimondo** [comment]
I fear I am not qualified to answer those questions since I'm really just experimenting with things by try and error. One thing I am definitely missing in Torch is the ability to run the forward and backward passes only up to a certain layer, like Caffe can do. That way it would be much easier to apply some custom manipulations in the middle, even use conditions, and then continue the forward pass from there. Not sure, maybe there is a way to do this in Torch, too, but so far I didn't find out how to do that.

So I don't know how a swap operation would work in the backward pass.

> indexNode, node are internal variable in ipairs(netG.forwardnodes)?How to use that code? insert that code after netG:forward(X)?

Yes exactly, it has to run after netG:forward(X). Of course you can change the code to extract other layers or types of layers. That's another annoying thing about Torch - I did not find out yet if it is possble to give layers a unqiue ID since that would make it so much easier to access them instead of having to iterate over than and identify them by index and type.

> By the way ,do you know how to remove a layer in gmodule?such as net_G:remove(i),but gmodule have not remove() function.

I guess you are talking about removing a layer from a pre-trained module? The way I'd do that is to first generate a new module that does not contain that layer (and has all the correct connections) and then copy the weights and biases from the loaded old model into the corresponding layers of the new one.

**Comment by royal-feng** [comment]
1.Actually, I don't need a swap operation work in the backward pass. The function of swap module like a feature swap in a certain middle features. 
2.I remove a layer from a pre-trained model like what you describe. 

I am interesting in the inpainting work, such as subtract some region in a image, then to fill in it.
![4](https://cloud.githubusercontent.com/assets/14287863/24301726/4ca17468-10eb-11e7-9c1a-a88068b748a7.jpg)
![4](https://cloud.githubusercontent.com/assets/14287863/24301786/7423ecc8-10eb-11e7-911a-bcfbaa956c60.jpg)
![4](https://cloud.githubusercontent.com/assets/14287863/24301794/79f2c30e-10eb-11e7-8f22-e7af1ea317ab.jpg)

I use the swap module to swap masked region feature with non-masked region feature
Do you have some suggestions? 

I am a new to use torch,learn so much from you. By the way, when you met some problem in torch ,where you would ask, I mean any "torch forum" or look up the torch document?
